sklearn/ensemble/_hist_gradient_boosting/grower.py

```diff
@@ -39,7 +39,7 @@ class TreeNode:
     ----------
     depth : int
         The depth of the node, i.e. its distance from the root.
-    sample_indices : ndarray of shape (n_samples_at_node,), dtype=np.uint
+    sample_indices : ndarray of shape (n_samples_at_node,), dtype=np.uint32
```

Comment:
Are you sure this is not `np.uint64` instead?

On my macOS laptop:

```python
>>> np.uint
<class 'numpy.uint64'>
```

Reply:
I'm not completely sure, but I followed implementation and `sample_indices` parameter is [provided by splitter](https://github.com/scikit-learn/scikit-learn/blob/6396dcb450e1ba5f897517796432c6de9fb709f0/sklearn/ensemble/_hist_gradient_boosting/grower.py#L393). Splitter's partition [has np.int32 dtype](https://github.com/scikit-learn/scikit-learn/blob/6396dcb450e1ba5f897517796432c6de9fb709f0/sklearn/ensemble/_hist_gradient_boosting/splitting.pyx#L217). Therefore I decided to set it as `np.uint32` but I can change it to `np.int64`.

Reply:
I think `uint32` is correct, see https://github.com/scikit-learn/scikit-learn/blob/206769cc5794fbe1dc1d2579f33843f4c87b1e4b/sklearn/ensemble/_hist_gradient_boosting/histogram.pyx#L105-L117.

(Question to myself: Shouldn't it be `intp`??)

Reply:
> (Question to myself: Shouldn't it be intp??)

In general, I think we should be careful to not make the dtype of fitted estimator attributes depend on the bitness of the Python runtime environment used for training the model, otherwise the resulting pickling files might be complex to load in another environment.

E.g. one could imagine training a model on a amd64 or aarch64 server and then deploying the model for prediction in a 32 bit WASM environment (on the "edge").

However in this particular case, the `TreeNode` class of the grower module should be transient and only be used at training time, so we are free to use a potentially runtime-specific datatype. However we need to double check that this choice of datatype does not impact the prediction datastructures stored as fitted attributes on the estimator object.

Reply:
About `uint32` vs `uint64`: any would do. I don't anticipate training an HGBDT model on a more than 4B training set any time soon.

Reply:
@ogrisel Enlightening answers. Could we please document those in the dev docs?